log.hpp

```diff
@@ -0,0 +1,58 @@
+
+#pragma once
+
+#include <utility>
+#include <cstdio>
+#include <cstdlib>
+#include "panic.hpp"
+
+inline void single_print(const char * arg, const char * next)
+{
+    panic(std::printf("%s%s", arg, next) < 0);
```

Comment:
You can just `assert` instead of `panic()`. Why introduce a new thing for one to look up? :)
But in this case, we can safely remove it. It's not that a `printf` can never fail. It's that we don't care that much.